### ISSUE TITLE
Block level marks in combination with decorations causes NodeViews to be reconstructed

### ISSUE THREAD

**Issue by superchu** [ProseMirror/prosemirror#872]
### Issue details

There seems to be an issue with block level marks in combination with decorations. When a decoration is rendered next to a node that has a block level mark, the nodeview for that node gets reconstructed when the decoration is removed.

### Steps to reproduce

I've created a simple example to demonstrate this. In the linked example we have a basic schema that allows a link-mark on images. There's also a custom plugin that renders a decoration when a image-node is selected + a custom node view for images where I just do a `console.log` when it gets destroyed.

Also, to make the issue more obvious I'm adding a timestamp to the src-attribute of images when it renders.

https://glitch.com/edit/#!/remix/beneficial-wash

### ProseMirror version

Latest

### Affected platforms

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Internet Explorer (not tested)

### Screenshots / Screencast (Optional)

![pm](https://user-images.githubusercontent.com/1313236/48927151-f3550b00-ef27-11e8-9193-29c26c35fc15.gif)

**Comment by marijnh** [comment]
Note that the editor is allowed to reconstruct any node it wants at any time—you can't depend on your node views having a given lifetime.

That being said, the reconstruction in this case seems unnecessary, and it'd be interesting to figure out why it is happening.

**Comment by superchu** [comment]
Yep, I'm aware of that - but as you said, in this case it seemed like there's no obvious reason for it which is why I thought it'd be a good idea to raise it and get your thoughts :)

**Comment by superchu** [comment]
Been trying to debug this a bit more. Here's some findings:

In `updateChildren` there's a call to `iterDeco` (https://github.com/ProseMirror/prosemirror-view/blob/master/src/viewdesc.js#L587). Inside that function `parent.childCount` is used for a for-loop. `childCount` does not take the block mark into account, so in our example this number is `6`.

In the `onNode` callback for each child, `findNodeMatch` is called and inside that function `this.top.children` (https://github.com/ProseMirror/prosemirror-view/blob/master/src/viewdesc.js#L963) is used as a reference for matching. That's the array of all childviews, which also includes the block mark, so in our example the length of this array is `7`.

Basically what ends up happening is that `destroyBetween` is called with `start: 2` and `end: 3` for the image node without a link mark, where `2` is the index of the decoration and `3` is the index of the image node.

For the image node with a link mark, `destroyBetween` is called with `start: 5` and `end: 7`, where `5` is the index of the decoration, `6` the link mark, and `7` the image node.

I'm guessing the issue is that `findNodeMatch` is never called for the link mark itself, so `this.index` is not increased.

Hope this helps a bit!

**Comment by marijnh** [comment]
Attached patch helps with this. Again, reconciling the DOM with the document is a heuristic process, and there's no guarantees made of any kind for what the library's algorithm will and will not match, but this was a reasonably common case (it'd also occur when deleting a node in front of an inline mark), so addressing it seems like a good idea.
